Make the package hash consistent

tgamblin opened the pull request:
This PR ensures that the package hash is consistent across Python versions, and adds a lot of tests to ensure that.  It also makes the package hash less sensitive, to avoid superfluous rebuilds in CI.

Now that we have #25310 (reusing installed packages), we want to make Spack hashes more fine-grained to preserve provenance.  To do that we want to switch to using full hashes everywhere.  Full hashes include all build dependencies, a hash of the canonicalized package file, and the hashes of artifacts and patches used in the build.

The existing full hash worked by editing the AST of a `package.py` file, removing multimethods (where statically possible b/c we know they won't be triggered), and stripping comments, docstrings, and directives.  Comments and docstrings don't affect the build, and directives are represented by other metadata on the spec.  Once all these modifications were done, we dumped the `repr()` of the AST and hashed that.

There were a few problems with the old package hash:
1. The AST changes pretty frequently from Python version to Python version, and so does its `repr()`.  So, Spack running under Python 2.7, 3.5, and 3.8 may calculate entirely different full hashes for the same `package.py` file.  This isn't as big a deal *right now*, but it needs to be fixed if we are going to move to full hashes.
2. Static multimethod reosolution (removing multi-methods if we know they won't be used) was not working in all cases in the previous version of the hash, and there weren't a lot of tests for it.  This resulted in the old package hash stripping *any* function with a decorator from the canonical representation, so we would lose some functions that affected builds.
3. The old docstring remover would only strip the first free-standing string in a module/class/method.  We want to strip all of them (a few packages had extra strings lying around).
4. The old package hash would not remove directive nested in complex logic at the package level (`if`, `for`, `with when()`, etc.), and that constitutes a lot of logic for some packages, even when all the loops are doing is setting up complex directives.  
5. The old package hash was considering changes to `maintainers` and `tags`, and we really don't want those to require a rebuild.  

This PR makes the following modifications:

- [x] To solve the AST issue, add `spack.util.unparse`, which regenerates Python source code from a parsed AST.  This is based on https://github.com/simonpercivall/astunparse, which is consistent across Python 2.7-3.8, with back ports from the new `ast.unparse` function in Python 3.9+.
    * The `unparse()` method implemented here is consistent for Python 2.7 and 3.5-3.10.
    * We added a `py_ver_consistent=True` argument to `unparse()`, which handles the following cases:
         * Ensure that `*args` and `**kwargs` are last in function calls, for consistency with the Python 2 AST (which has no idea what position they're in)
         * Always unparse `print()` as a function (Python 2's AST treats it as its own statement) and do *not* differentiate between `print(a, b, c)` and `print((a, b, c))`.  These have different semantics in Python 2 and Python 3, but they almost certainly do not affect the results of builds, so switching between them won't affect the package hash.
         * Unparse unicode literals the way Python 2 would.  Basically, this means that all non-ASCII characters will use `\x` encoding in output, and unicode literals will be unparsed with a `u` prefix.
    * Our `unparse()` is consistent with `ast.unparse()` when `py_ver_consistent` is not used, so that we *may* be able to switch to the builtin `ast.unparse()` one day (if we ever only support Python 3.9+).  This is done with backports of things like `ast.unparse`'s parenthesis precedence algorithm.
    * Add some tests that guarantee that packages will have the same hash across different python versions.
- [x] Fix multimethod resolution in `package_hash.py` to preserve non-`when` decorators, and handle dynamic `@when` statements better. Add a bunch of tests for this.
- [x] Fix docstring stripping to remove any free-standing string in a `package.py` file.  This is safe as these strings are used as comments -- they don't affect build results.
- [x] Add some new commands to deal with package hashes:
    * `spack pkg source <spec>`: dump source for a `package.py` file
    * `spack pkg source --canonical <spec>`: dump canonical source for a `package.py` file, with comments and docstrings stripped and multimethods removed.
    * `spack pkg hash <spec>`: print the canonical package hash for the spec.
- [x] Remove the old AST-dump-based `package_hash` and use the new, stable `unparse`-based one.
- [x] New unparser descends into control logic a the package level and removes anything (loops, ifs, with statements) that is comprised only of directives.  This makes the package hash less sensitive.
- [x] Added more attributes for removal (`maintainers`, `tags`) to make the hash less sensitive.

I've verified that this is consistent for Python 2.7 and 3.5-3.10 across all Spack packages by unparsing them and comparing their canonicalized source.  The test packages in `lib/spack/spack/test/data/unparse` and the `test_package_hash_consistency()` test are meant to preserve this with some spot checks of interesting cases.

This should trigger a large CI rebuild on *this PR*, but it may reduce CI churn in the future (though I think CI was only using one version of Python, so maybe not).

tgamblin committed:
externals: add astunparse

Add `astunparse` as `spack_astunparse`. This library unparses Python ASTs and we're
adding it under our own name so that we can make modifications to it.

Ultimately this will be used to make `package_hash` consistent across Python versions.

tgamblin committed:
unparse: Make unparsing consistent for 2.7 and 3.5-3.10

Previously, there were differences in the unparsed code for Python 2.7 and for 3.5-3.10.
This makes unparsed code the same across these Python versions by:

    1. Ensuring there are no spaces between unary operators and
       their operands.
    2. Ensuring that *args and **kwargs are always the last arguments,
       regardless of the python version.
    3. Always unparsing print as a function.
    4. Not putting an extra comma after Python 2 class definitions.

Without these changes, the same source can generate different code for different
Python versions, depending on subtle AST differences.

One place where single source will generate an inconsistent AST is with
multi-argument print statements, e.g.:

```
    print("foo", "bar", "baz")
```

In Python 2, this prints a tuple; in Python 3, it is the print function with
multiple arguments.  Use `from __future__ import print_function` to avoid
this inconsistency.

tgamblin committed:
unparser: add block() context manager for indentation

This is a backport of a refactor from cpython 3.9

tgamblin committed:
unparser: refactor delimiting with context managers in ast.unparse

Backport of https://github.com/python/cpython/commit/4b3b1226e86df6cd45e921c8f2ad23c3639c43b2

tgamblin committed:
unparser: implement operator precedence algorithm for unparser

Backport operator precedence algorithm from here:
    https://github.com/python/cpython/commit/397b96f6d7a89f778ebc0591e32216a8183fe667

This eliminates unnecessary parentheses from our unparsed output and makes Spack's unparser
consistent with the one in upstream Python 3.9+, with one exception.

Our parser normalizes argument order when `py_ver_consistent` is set, so that star arguments
in function calls come last.  We have to do this because Python 2's AST doesn't have information
about their actual order.

If we ever support only Python 3.9 and higher, we can easily switch over to `ast.unparse`, as
the unparsing is consistent except for this detail (modulo future changes to `ast.unparse`)

tgamblin committed:
package_hash: rework `RemoveDirectives` and add a test

Previously we used `directives.__all__` to get directive names, but it wasn't
quite right -- it included `DirectiveMeta`, etc.  It's not wrong, but it's also
not the clearest way to do this.

- [x] Refactor `@directive` to track names in `directive_names` global
- [x] Rename `_directive_names` to `_directive_dict_names` in `DirectiveMeta`
- [x] Add a test for `RemoveDirectives`

Co-authored-by: Danny McClanahan <1305167+cosmicexplorer@users.noreply.github.com>

tgamblin committed:
unparser: handle unicode string literals consistently across Python versions

Python 2 and 3 represent string literals differently in the AST. Python 2 requires '\x'
literals, and Python 3 source is always unicode, and allows unicode to be written
directly.  These also unparse differently by default.

- [x] modify unparser to write both out the way `repr` would in Python 2 when
  `py_ver_consistent` is provided.

tgamblin committed:
package_hash: remove all unassigned strings, not just docstrings

Some packages use top-level unassigned strings instead of comments, either just after a
docstring on in the body somewhere else. Ignore those strings becasue they have no
effect on package behavior.

- [x] adjust RemoveDocstrings to remove all free-standing strings.
- [x] move tests for util/package_hash.py to test/util/package_hash.py

Co-authored-by: Danny McClanahan <1305167+cosmicexplorer@users.noreply.github.com>

tgamblin committed:
package_hash: fix handling of multimethods and add tests

Package hashing was not properly handling multimethods. In particular, it was removing
any functions that had decorators from the output, so we'd miss things like
`@run_after("install")`, etc.

There were also problems with handling multiple `@when`'s in a single file, and with
handling `@when` functions that *had* to be evaluated dynamically.

- [x] Rework static `@when` resolution for package hash
- [x] Ensure that functions with decorators are not removed from output
- [x] Add tests for many different @when scenarios (multiple @when's,
      combining with other decorators, default/no default, etc.)

Co-authored-by: Danny McClanahan <1305167+cosmicexplorer@users.noreply.github.com>

tgamblin committed:
package_hash: add test to ensure consistency across Python versions

Our package hash is supposed to be consistent from python version to python version.
Test this by adding some known unparse inputs and ensuring that they always have the
same canonical hash.  This test relies on the fact that we run Spack's unit tests
across many python versions.  We can't compute for several python versions within the
same test run so we precompute the hashes and check them in CI.

tgamblin committed:
package_hash: add test to ensure that every package in Spack can be unparsed

- [x] add option to canonical source to *not* filter multimethods
- [x] add test to unparse every package in builtin

tgamblin committed:
unparser: Don't put unnecessary parentheses on class declarations

Backport of
* https://github.com/python/cpython/commit/25160cdc4775a1ddb4e37c8bf5a6e31ad9c146ed

tgamblin committed:
unparser: Don't omit parenthesis when unparsing a slice

Backport of
  https://github.com/python/cpython/commit/c102a148256b00b7d48c51a1a97df19042e603de

Includes support for Python 2.

tgamblin committed:
package_hash: switch to using canonical source instead of AST repr

We are planning to switch to using full hashes for Spack specs, which means that the
package hash will be included in the deployment descriptor. This means we need a more
robust package hash than simply dumping the `repr` of the AST.

The AST repr that we previously used for package content is unreliable because it can
vary between python versions (Python's AST actually changes fairly frequently).

- [x] change `package_hash`, `package_ast`, and `canonical_source` to accept a string for
      alternate source instead of a filename.
- [x] consolidate package hash tests in `test/util/package_hash.py`.
- [x] remove old `package_content` method.
- [x] make `package_hash` do what `canonical_source_hash` was doing before.
- [x] modify `content_hash` in `package.py` to use the new `package_hash` function.

Co-authored-by: Danny McClanahan <1305167+cosmicexplorer@users.noreply.github.com>

tgamblin committed:
commands: add `spack pkg source` and `spack pkg hash`

To make it easier to see how package hashes change and how they are computed, add two
commands:

* `spack pkg source <spec>`: dumps source code for a package to the terminal

* `spack pkg source --canonical <spec>`: dumps canonicalized source code for a
   package to the terminal. It strips comments, directives, and known-unused
   multimethods from the package. It is used to generate package hashes.

* `spack pkg hash <spec>`: This gives the package hash for a particular spec.
  It is generated from the canonical source code for the spec.

- [x] `add spack pkg source` and `spack pkg hash`
- [x] add tests
- [x] fix bug in multimethod resolution with boolean `@when` values

Co-authored-by: Greg Becker <becker33@llnl.gov>

tgamblin committed:
unparser: treat `print(a, b, c)` and `print((a, b, c))` the same

We can't tell `print(a, b, c)` and `print((a, b, c))` apart -- both of these expressions
generate different ASTs in Python 2 and Python 3.  However, we can decide that we don't
care.  This commit treats both of them the same when `py_ver_consistent` is set with
`unparse()`.

This means that the package hash won't notice changes from printing a tuple to printing
multiple values, but we don't care, because this is extremely unlikely to affect the build.
More than likely this is just an error message for the user of the package.

- [x] treat `print(a, b, c)` and `print((a, b, c))` the same in py2 and py3
- [x] add another package parsing test -- legion -- that exercises this feature

scheibelp commented on lib/spack/external/spack_astunparse/unparser.py:
```diff
@@ -74,14 +84,22 @@ def write(self, text):
-        self._indent += 1
+    class _Block:
+        """A context manager for preparing the source for blocks. It adds
+        the character':', increases the indentation on enter and decreases
```
(minor) this is a typo (no space between "character" and "':'". More importantly, I'm curious if this comment will be displayed in the PR (I made this comment on 493c867)

cosmicexplorer replied:
I can read it when looking at the PR "conversation" tab, so I believe the answer is yes!

becker33 commented on lib/spack/spack/cmd/pkg.py:
```diff
@@ -201,14 +212,49 @@ def pkg_changed(args):
+def pkg_hash(args):
+    """dump canonical source code hash for a package spec"""
+    specs = spack.cmd.parse_specs(args.spec, concretize=False)
+    if len(specs) != 1:
+        tty.die("spack pkg hash requires exactly one spec")
```
Why make this restriction? Couldn't we just print each hash sequentially on its own line?

tgamblin replied:
Yep good point.  Fixed.

becker33 commented on lib/spack/spack/test/data/unparse/amdfftw.txt:
```diff
@@ -0,0 +1,259 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
```
For all of these packages, can you replace the docstring description with a description of what about this package is particularly challenging for the unparser or what this package is testing about the unparser?

tgamblin replied:
done

becker33 commented on lib/spack/spack/test/util/package_hash.py:
```diff
@@ -0,0 +1,360 @@
+        for i in range(10):
+            print(i)
+
+    def method1():
```
Are these two methods supposed to have the same name? If so it could use a comment explaining why.

tgamblin replied:
it actually doesn't matter for this test but they don't need to be named the same thing and probably shouldn't be.  I renamed them method1 and method2 and added a check to ensure that they're still in the unparsed output.

becker33 commented on lib/spack/spack/test/util/package_hash.py:
```diff
@@ -0,0 +1,360 @@
+    assert "FOUR" in filtered
+    assert "FIVE" in filtered
+
+    # we don't know if default or THREE will win, include both
+    filtered = ph.canonical_source("pkg@2.0", source=many_multimethods)
+    assert "ONE" in filtered
+    assert "TWO" not in filtered
+    assert "THREE" in filtered
+    assert "FOUR" not in filtered
+    assert "FIVE" in filtered
```
Can you explain a bit why this is the case? I don't think I can adequately review this block without a quick chat about why it is the way it is.

tgamblin replied:
Is the [explanation in package_hash.py](https://github.com/spack/spack/pull/28156/files#diff-c0e9963b0d8dcee67b8fc64f623170a84ac7c13892f504c95106e8d1c4cdaf13R136-R178) sufficient?

The reason you need both is because you don't know until runtime whether all the `@when`'s on  `THREE` will end up being true, b/c you don't know whether you'll end up running the build on Darwin or not.  If you're on Darwin, THREE wins.  otherwise default does b/c TWO, FOUR, and FIVE are false, and the default is preserved in case nothing wins.

becker33 replied:
Yes, I think that's sufficient. It's frustrating that we can't fully canonicalize -- I suspect that most of these packages could have `'platform=darwin'` instead of `sys.platform == 'darwin'`, which would let us trim a little more.

tgamblin replied:
While I was thinking about this, I realized that we should consider disallowing dynamic conditions like this in the future for another reason: it interferes with indexing the package metadata. We can deal with that later, too.

cosmicexplorer commented on lib/spack/spack/test/data/unparse/README.md:
```diff
@@ -0,0 +1,6 @@
+
+These are test packages for testing Spack's unparser.
+
+They are named with `.txt` suffixes so that they're not considered
+proper source files by the various checkers used in Spack CI.
```
This seems like a good workaround. Is there any documentation about spack checkers we might want to add this as a note to?

cosmicexplorer replied:
Do I understand correctly that the content of these files can be regenerated at any time by calling `spack pkg source <package>`? Could we add a small note on that in this file so someone who wants to add an additional test package later gets how to do that?

tgamblin replied:
They're just copies of packages -- they need to remain static so their hashes don't change.

cosmicexplorer replied:
Ok, that sounds reasonable!

cosmicexplorer commented on lib/spack/spack/package.py:
```diff
@@ -52,6 +52,7 @@
 import spack.store
 import spack.url
 import spack.util.environment
+import spack.util.hash
```
Why is this imported here if the rest of the changes in this file do not use that import? Is it getting magically invoked somehow?

tgamblin replied:
I am very confused by this -- it seems like flake8 is overlooking it.

tgamblin replied:
I just removed this.

cosmicexplorer commented on lib/spack/spack/directives.py:
```diff
@@ -730,3 +733,7 @@ class DependencyPatchError(DirectiveError):
+
+
+#: add all directive names to __all__
+__all__.extend(directive_names)
```
This is a good change!

cosmicexplorer commented on lib/spack/spack/directives.py:
```diff
@@ -59,6 +59,9 @@ class OpenMpi(Package):
 #: These are variant names used by Spack internally; packages can't use them
 reserved_names = ['patches', 'dev_path']
 
+#: Names of possible directives
```
```suggestion
#: Names of possible directives. This list is populated elsewhere in the file and then added to `__all__` at the bottom.
```

cosmicexplorer commented on lib/spack/spack/cmd/pkg.py:
```diff
@@ -201,14 +212,49 @@ def pkg_changed(args):
+    spec = specs[0]
+
+    # regular source dump -- just get the package and print its contents
+    if not args.canonical:
```
Recommend flipping the logical condition here since it can sometimes be confusing to read `if not x` when both sides of `if`/`else` are used. This might look like:

```python
    if args.canonical:
        content = ph.canonical_source(spec)
        sys.stdout.write(content)
        return

    # regular source dump -- just get the package and print its contents
    filename = spack.repo.patch.filename_for_package_name(spec.name)
    if sys.stdout.isatty():
        tty.msg("%s:" % filename)
    with open(filename) as f:
        sys.stdout.write(f.read())
```

However, it's also not clear to me why we would only `tty.msg()` the `filename` when we do a non-`canonical` source dump. To remove some of this logic, I might recommend the following: 

```python
    filename = spack.repo.path.filename_for_package_name(spec.name)
    if sys.stdout.isatty():
        tty.msg("%s:" % filename)

    if args.canonical:
        content = ph.canonical_source(spec)
    else:
        with open(filename) as f:
            content = f.read() # we might have an llnl.util.filesystem method for this

    sys.stdout.write(content)
```

tgamblin replied:
ok did this.

cosmicexplorer commented on lib/spack/spack/test/cmd/pkg.py:
```diff
@@ -236,3 +236,51 @@ def test_pkg_fails_when_not_git_repo(monkeypatch):
+    assert "Check that multimethods work with boolean values" in source
+
+    canonical_1 = pkg("source", "--canonical", "multimethod@1.0")
+    assert "@when" not in canonical_1
```
I get worried when I see `assert ... not in ...`, since `not in` could be true if it spit out the completely wrong info as well. Would it be too brittle if we took the output of the `spack pkg source --canonical multimethod@1.0` command as another `.txt` file and simply compared it directly to the output here?

tgamblin replied:
I think that's pretty brittle, and there are a lot of `in` tests in addition to `not in` tests here.

cosmicexplorer replied:
Ok great!

cosmicexplorer commented on lib/spack/spack/test/cmd/pkg.py:
```diff
@@ -236,3 +236,51 @@ def test_pkg_fails_when_not_git_repo(monkeypatch):
+
+
+def test_pkg_hash(mock_packages):
+    pkg("hash", "multimethod")
```
Is this just testing that the `spack pkg hash multimethod` command doesn't fail? Could we test against its actual hash now that the hash is consistent?

tgamblin replied:
The hash is only consistent if the package doesn't change.  So testing that would require another argument to grab the source from some alternative and unchanging place, but we already have other tests that do that... I didn't think it was worth the trouble.

cosmicexplorer replied:
Ok, great!

cosmicexplorer commented on lib/spack/spack/test/package_sanity.py:
```diff
@@ -56,6 +57,19 @@ def test_packages_are_pickleable():
+        except Exception:
+            failed_to_unparse.append(name)
+
+    if failed_to_unparse:
+        tty.msg('The following packages failed to unparse: ' +
+                ', '.join(failed_to_unparse))
```
Is this unfinished code? I would assume we'd use an `assert` here in order to make sure this test fails?

cosmicexplorer commented on lib/spack/spack/util/package_hash.py:
```diff
@@ -4,23 +4,30 @@
-    """Transformer that removes docstrings from a Python AST."""
+    """Transformer that removes docstrings from a Python AST.
+
+    This removes *all* unassigned strings from the body of functions, classes,
```
```suggestion
    This removes *all* strings that aren't on the RHS of an assignment statement from the body of functions, classes,
```

cosmicexplorer commented on lib/spack/spack/util/package_hash.py:
```diff
@@ -74,82 +81,206 @@ def visit_ClassDef(self, node):  # noqa
+    whether we're on the darwin platform *at build time* (the darwin condition for
+    implementation 3 is dynamic). We know the package spec statically; we don't know
+    statically what the runtime environment will be. We need to include things that can
+    possible affect package behavior in the package hash, and we want to exclude things
```
```suggestion
    possibly affect package behavior in the package hash, and we want to exclude things
```

cosmicexplorer replied:
This is a fantastically helpful explanation.

scottwittenburg replied:
I agree, this docstring really helped me understand the issue with multimethods.

cosmicexplorer commented on lib/spack/spack/util/package_hash.py:
```diff
@@ -74,82 +81,206 @@ def visit_ClassDef(self, node):  # noqa
-        return None
+    def visit_FunctionDef(self, func):  # noqa
+        # if the function def wasn't visited on the first traversal there is a problem
+        assert func.name in self.methods, "Inconsistent package traversal!"
```
Recommend turning this into an `Exception` subclass before merging.

tgamblin replied:
Why?  If this isn't true, it's a logic error that should come out in testing, not something that should be dealt with at runtime by any client code.

cosmicexplorer replied:
Ok, I misunderstood the intent of "there is a problem" in the comment! I think this is fine as-is then.

cosmicexplorer commented on lib/spack/spack/util/package_hash.py:
```diff
@@ -74,82 +81,206 @@ def visit_ClassDef(self, node):  # noqa
+def package_hash(spec, source=None):
+    """Get a hash of a package's canonical source code.
+
+    This is the has function used to determine whether a spec needs a rebuild when a
```
```suggestion
    This function is used to determine whether a spec needs a rebuild when a
```

cosmicexplorer commented on lib/spack/spack/util/unparse/unparser.py:
```diff
@@ -0,0 +1,1170 @@
+from six import StringIO
+
+
+# TODO: if we require Python 3.7, use its `nullcontext()`
```
Possibly a job for #21720.

tgamblin replied:
yeah we should probably resurrect that.

cosmicexplorer replied:
you got it boss

cosmicexplorer commented on lib/spack/spack/util/unparse/unparser.py:
```diff
@@ -0,0 +1,1170 @@
+    yield
+
+
+# Large float and imaginary literals get turned into infinities in the AST.
```
Is there documentation for this behavior in the `ast` or `unparse` modules anywhere?

tgamblin replied:
No, but unparser.py is basically an external -- I'm hesitant to change much more about it.

cosmicexplorer replied:
Oh, right! Thanks--sounds great to keep as-is!

cosmicexplorer commented on lib/spack/spack/util/unparse/unparser.py:
```diff
@@ -0,0 +1,1170 @@
+            2. Always unparsing print as a function.
+            3. Unparsing Python3 unicode literals the way Python 2 would.
+
+        Without these changes, the same source can generate different code for Python 2
+        and Python 3, depending on subtle AST differences.  The first of these two
+        causes this module to behave differently from Python 3.8+'s `ast.unparse()`
```
Good docstring!

cosmicexplorer reviewed (dismissed):
This is awesome!!!!!!! I'm marking approve because I would accept this as-is, but will pay attention to any discussion on my review comments.

becker33 commented on lib/spack/spack/util/package_hash.py:
```diff
@@ -74,82 +83,206 @@ def visit_ClassDef(self, node):  # noqa
+                    # unless some other constraint is False.
+                    conditions.append(None)
+
+        # anything defined without conditions will overwrite prior definitions
+        if not conditions:
+            self.methods[func.name] = []
+
+        # add all discovered conditions on this node to the node list
+        impl_conditions = self.methods.setdefault(func.name, [])
+        impl_conditions.append((func, conditions))
+
+        # don't modify the AST -- return the untouched function node
+        return func
```
I'm a little concerned with how this will play with multimethods on inherited packages. Can you make sure to add a test for parsing a package that inherits from a package with multimethod and overrides, and another for adding a condition to the multimethod?

tgamblin replied:
So, at the moment there are no inherited multimethods in Spack mainline (either in `build_systems` or in any package that's imported from another), and base packages aren't included in the package hash at all.

There is a question of where to draw the line for the package hash.  I think eventually we should consider making specs even finer-grained and include the spack version or commit in their hashes.  At the same time we might want to include all bases up to but not including `PackageBase`, but then that overlaps with the spack hash in terms of what's represented in a spec.  We could include anything that's *not* in Spack itself, I suppose.

The multimethod stuff is separate because whether we preserve subclass multi methods can depend on whether we're overriding a multimethod from a parent class.  I think that analysis should get factored in, but currently we don't have that information in the AST analysis -- we would need to look up the parent class and parse its AST, as well.

So ultimately I think it's important to handle all this, but I think we should do it in a separate PR.

becker33 replied:
I think it's okay to push this to another PR, but that does mean we will need to have another hash-breaking PR in the future.

tgamblin replied:
I think this becomes increasingly less impactful the sooner we make reuse the default.

becker33 replied:
It will still cause a CI backlog at least, even if users are insulated from it

becker33 commented on lib/spack/spack/util/package_hash.py:
```diff
@@ -74,82 +83,206 @@ def visit_ClassDef(self, node):  # noqa
+    root = ast.parse(source)
     root = RemoveDocstrings().visit(root)
-
     RemoveDirectives(spec).visit(root)
 
-    fmm = TagMultiMethods(spec)
-    fmm.visit(root)
+    if filter_multimethods:
+        tagger = TagMultiMethods(spec)
+        tagger.visit(root)
+        root = ResolveMultiMethods(tagger.methods).visit(root)
```
Can we clean up the calling convention for these? It looks like `RemoveDirectives.visit` and `TagMultiMethods.visit` operate in place, but `RemoveDocstrings.visit` and `ResolveMultiMethods.visit` return a copy

tgamblin replied:
Yep this is a good idea.

tgamblin replied:
The tagger is just a visitor and doesn't modify the tree, so I didn't tweak that one, but the other two should be assignments because they are `NodeTransformer` subclasses.  I added some comments to make this clearer

tgamblin commented:
@becker33 this is ready for another look, and I have some questions.

tgamblin commented:
I guess the last question is what to do with `unparser.py`.  Currently its coverage is low because we're not running the full unparse test suite on it.  But, it's heavily modified and has code from `astunparse` and a bunch of backports from upstream Python.

I think maybe it should be its own repo in the spack org, with its own tests to ensure that it is consistent with upstream `ast.unparse()`.  That would probably let us test it more easily, and it would let us monitor its coverage more easily.  I think other projects might benefit from a canonical unparser, too.  Any thoughts on this?

tgamblin committed:
unparser: do a better job of roundtripping strings

Handle complex f-strings.  Backport of:

    https://github.com/python/cpython/commit/a993e901ebe60c38d46ecb31f771d0b4a206828c#

tgamblin committed:
unparser: rename `_Class()` methods to `visit_Class()` to mirror upstream

These are refactors that have happened in upstream `ast.unparse()`

tgamblin committed:
unparser: rename `t` to `node` to mirror upstream

These refactors have happened in upstream `ast.unparse()`

tgamblin committed:
unparser: add unparser unit tests

These are the unit tests from astunparse, converted to pytest, with a few backports from
upstream cpython. These should hopefully keep `unparser.py` well covered as we change it.

tgamblin committed:
unparser: handle package-level loops, if statements, and with blocks

Many packages implement logic at the class level to handle complex dependencies and
conflicts. Others have started using `with when("@1.0"):` blocks since we added that
capability. The loops and other control logic can cause some pure directive logic not to
be removed by our package hashing logic -- and in many cases that's a lot of code that
will cause unnecessary rebuilds.

This commit changes the unparser so that it will descend into these blocks. Specifically:

  1. Descend into loops, if statements, and with blocks at the class level.
  2. Don't look inside function definitions (in or outside a class).
  3. Don't look at nested class definitions (they don't have directives)
  4. Add logic to *remove* empty loops/with blocks/if statements if all directives
     in them were removed.

This allows our package hash to ignore a lot of pure metadata that it was not ignoring
before, and makes it less sensitive.

In addition, we add `maintainers` and `tags` to the list of metadata attributes that
Spack should remove from packages when constructing canonoical source for a package
hash.

- [x] Make unparser handle if/for/while/with at class level.
- [x] Add tests for control logic removal.
- [x] Add a test to ensure that all packages are not only unparseable, but also
      that their canonical source is still compilable. This is a test for
      our control logic removal.
- [x] Add another unparse test package that has complex logic.

tgamblin commented:
For now I just added unit tests to ensure that we're covering all of the unparser logic.  I think we can factor this out later.

I looked at packages and discovered that we weren't handling `with when()`, `for`, `if`, etc. or some attributes like `maintainers` properly.  I added `maintainers` and `tags` to ignored attributes, and the unparser now also removes nested logic if it contains only directives.  I think this is a much better package hash now.

@becker33 can you take one more look?

scottwittenburg reviewed (commented):
I won't up or down this, but I did look through it deeply enough to gain a better understanding of how the package was computed before, and why this is an improvement.  Thanks for the detailed PR description and some great docstrings discussing the finer points.

As you pointed out somewhere, each gitlab CI pipeline stack is only run with a single version of python, which is baked into the container used to concretize the stack, so maybe this won't reduce the number of rebuilds for that reason.  But more consistent package hashing seems like an important prerequisite to using the full hash everywhere.

becker33 commented on lib/spack/spack/test/cmd/pkg.py:
```diff
@@ -236,3 +236,63 @@ def test_pkg_fails_when_not_git_repo(monkeypatch):
+
+def test_pkg_hash(mock_packages):
+    output = pkg("hash", "a", "b").strip().split()
+    assert len(output) == 2 and (len(elt) == 32 for elt in output)
```
```suggestion
    assert len(output) == 2 and all(len(elt) == 32 for elt in output)
```
Otherwise python will recognize that the generator is of non-zero length, and therefore treat it as true-ish even if all elements are `False`.

becker33 commented on lib/spack/spack/test/cmd/pkg.py:
```diff
@@ -236,3 +236,63 @@ def test_pkg_fails_when_not_git_repo(monkeypatch):
+    assert len(output) == 2 and (len(elt) == 32 for elt in output)
+
+    output = pkg("hash", "multimethod").strip().split()
+    assert len(output) == 1 and (len(elt) == 32 for elt in output)
```
```suggestion
    assert len(output) == 1 and all(len(elt) == 32 for elt in output)
```

becker33 reviewed (dismissed):
I think this has addressed all my review comments, excepting the issue we pushed to a future PR regarding multimethods inherited from another package in the MRO.

Two very minor issues with the tests, should be able to just commit the suggestions below.

tgamblin commented:
@becker33 all changes are committed and squashed into prior commits.  Thanks!

tgamblin commented:
@spackbot run pipeline

tgamblin commented:
@spackbot run pipeline
